File: expandablrecyclerview/src/main/java/com/umut/expandablrecyclerview/adapter/index/ExpandableIndexProvider.java
package com.umut.expandablrecyclerview.adapter.index;

import android.support.annotation.IntDef;
import android.support.annotation.Nullable;

import com.umut.expandablrecyclerview.adapter.ChildCoordinate;

import java.lang.annotation.Retention;
import java.lang.annotation.RetentionPolicy;

import static com.umut.expandablrecyclerview.adapter.index.ExpandableIndexProvider.ViewType.CHILD;
import static com.umut.expandablrecyclerview.adapter.index.ExpandableIndexProvider.ViewType.PARENT;

public interface ExpandableIndexProvider {

    int getTotalItemCount();

    int getAdapterIndexForParentAt(int adapterParentIndex);

    @ViewType
    int getViewType(int position);

    int parentIndexFromAdapterPosition(int adapterPosition);

    boolean isExpanded(int dataPosition);

    void onExpand(int dataPosition);

    void onCollapse(int dataPosition);

    @Nullable
    ChildCoordinate getChildCoordinateFromAdapterIndex(int adapterPosition);

    @IntDef({PARENT, CHILD})
    @Retention(RetentionPolicy.SOURCE)
    @interface ViewType {
        int PARENT = 0;
        int CHILD = 1;
    }
}

File: expandablrecyclerview/src/main/java/com/umut/expandablrecyclerview/adapter/index/ExpandableIndexProvider.kt
package com.umut.expandablrecyclerview.adapter.index

import androidx.annotation.IntDef
import com.umut.expandablrecyclerview.adapter.ChildCoordinate

interface ExpandableIndexProvider {
    companion object {
        const val PARENT = 0
        const val CHILD = 1

        const val UNSPECIFIED = -1
    }

    fun getTotalItemCount(): Int

    fun getAdapterIndexForParentAt(adapterParentIndex: Int): Int

    @ViewType
    fun getViewType(position: Int): Int

    fun parentIndexFromAdapterPosition(adapterPosition: Int): Int

    fun isExpanded(dataPosition: Int): Boolean

    fun onExpand(dataPosition: Int)

    fun onCollapse(dataPosition: Int)

    fun getChildCoordinateFromAdapterIndex(adapterPosition: Int): ChildCoordinate?

    @IntDef(PARENT, CHILD)
    @kotlin.annotation.Retention(AnnotationRetention.SOURCE)
    annotation class ViewType
}
